ast.ts

```diff
@@ -15,39 +20,57 @@ export type Class<A> = { a?: A, name: string, fields: Array<VarInit<A>>, methods
+  decls: Array<Scope<A>>,
+  inits: Array<VarInit<A>>, 
+  funs: Array<FunDef<A>>
+  body: Array<Stmt<A>>
+}
+
+export type Closure<A> = { 
+  a?: A, 
+  name: string, 
+  fields: Array<VarInit<A>>, 
+  apply: FunDef<A> 
+}
 
 export type Stmt<A> =
-  | {  a?: A, tag: "assign", name: string, value: Expr<A> }
-  | {  a?: A, tag: "return", value: Expr<A> }
-  | {  a?: A, tag: "expr", expr: Expr<A> }
-  | {  a?: A, tag: "if", cond: Expr<A>, thn: Array<Stmt<A>>, els: Array<Stmt<A>> }
-  | {  a?: A, tag: "while", cond: Expr<A>, body: Array<Stmt<A>> }
-  | {  a?: A, tag: "pass" }
-  | {  a?: A, tag: "field-assign", obj: Expr<A>, field: string, value: Expr<A> }
+  | { a?: A, tag: "assign", name: string, value: Expr<A> }
+  | { a?: A, tag: "return", value: Expr<A> }
+  | { a?: A, tag: "expr", expr: Expr<A> }
+  | { a?: A, tag: "if", cond: Expr<A>, thn: Array<Stmt<A>>, els: Array<Stmt<A>> }
+  | { a?: A, tag: "while", cond: Expr<A>, body: Array<Stmt<A>> }
+  | { a?: A, tag: "pass" }
+  | { a?: A, tag: "field-assign", obj: Expr<A>, field: string, value: Expr<A> }
 
 export type Expr<A> =
-    {  a?: A, tag: "literal", value: Literal }
-  | {  a?: A, tag: "id", name: string }
-  | {  a?: A, tag: "binop", op: BinOp, left: Expr<A>, right: Expr<A>}
-  | {  a?: A, tag: "uniop", op: UniOp, expr: Expr<A> }
-  | {  a?: A, tag: "builtin1", name: string, arg: Expr<A> }
-  | {  a?: A, tag: "builtin2", name: string, left: Expr<A>, right: Expr<A>}
-  | {  a?: A, tag: "call", name: string, arguments: Array<Expr<A>> } 
-  | {  a?: A, tag: "lookup", obj: Expr<A>, field: string }
-  | {  a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
-  | {  a?: A, tag: "construct", name: string }
+  | { a?: A, tag: "literal", value: Literal }
+  | { a?: A, tag: "id", name: string }
+  | { a?: A, tag: "binop", op: BinOp, left: Expr<A>, right: Expr<A> }
+  | { a?: A, tag: "uniop", op: UniOp, expr: Expr<A> }
+  | { a?: A, tag: "builtin1", name: string, arg: Expr<A> }
+  | { a?: A, tag: "builtin2", name: string, left: Expr<A>, right: Expr<A> }
+  | { a?: A, tag: "call", name: string, arguments: Array<Expr<A>> } 
+  | { a?: A, tag: "lookup", obj: Expr<A>, field: string }
+  | { a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
+  | { a?: A, tag: "construct", name: string }
+  | { a?: A, tag: "lambda", args: Array<string>, ret: Expr<A> }
 
 export type Literal = 
-    { tag: "num", value: number }
+  | { tag: "num", value: number }
   | { tag: "bool", value: boolean }
   | { tag: "none" }
 
 // TODO: should we split up arithmetic ops from bool ops?
-export enum BinOp { Plus, Minus, Mul, IDiv, Mod, Eq, Neq, Lte, Gte, Lt, Gt, Is, And, Or};
+export enum BinOp { Plus, Minus, Mul, IDiv, Mod, Eq, Neq, Lte, Gte, Lt, Gt, Is, And, Or };
 
 export enum UniOp { Neg, Not };
 
 export type Value =
-    Literal
+  | Literal
   | { tag: "object", name: string, address: number}
+  | { tag: "callable", name: string, address: number}
```

Comment:
What's the plan for the name of a `lambda`-created function?

Reply:
the lambda created function is compiled to wasm function named `$lambda_n` where `n` is a global counter in the environment. A callable value is created when a function or a lambda is compiled, where its `name` is the function name, and its `address` stores the table index. This callable value is assigned to variable of the function name for ordinary function, and assigned to the callable variable for lambda expression (e.g. `f = lambda x : 0`)